tests/TestCase/Consumption/LimitAttemptsExtensionTest.php

```diff
@@ -0,0 +1,146 @@
+
+        $extension->onResult($messageContext);
+
+        $this->assertSame(Processor::REJECT, $messageContext->getResult()->getStatus());
+        $this->assertSame('A copy of the message was sent with an incremented attempt count.', $messageContext->getResult()->getReason());
+    }
+
+    public function testResultShouldChangeToRejectIfMaxAttemptsIsExceeded()
+    {
+        $message = $this->createMessage(MaxAttemptsIsThreeJobStub::class);
+        $message->setProperty(LimitAttemptsExtension::ATTEMPTS_PROPERTY, 2);
+        $messageContext = $this->createMessageResult($message);
+
+        $this->assertSame(Processor::REQUEUE, $messageContext->getResult());
+
+        $extension = new LimitAttemptsExtension();
+
+        $extension->onResult($messageContext);
+
+        $this->assertSame(Processor::REJECT, $messageContext->getResult()->getStatus());
+        $this->assertSame('The maximum allowed attempts of 3 was exceeeded.', $messageContext->getResult()->getReason());
+    }
+
+    public function testResultShouldChangeToRejectIfGlobalMaxAttemptsIsExceeded()
+    {
+        $message = $this->createMessage(MaxAttemptsIsNotSetJobStub::class);
+        $message->setProperty(LimitAttemptsExtension::ATTEMPTS_PROPERTY, 2);
+        $messageContext = $this->createMessageResult($message);
+
+        $this->assertSame(Processor::REQUEUE, $messageContext->getResult());
+
+        $extension = new LimitAttemptsExtension(3);
+
+        $extension->onResult($messageContext);
+
+        $this->assertSame(Processor::REJECT, $messageContext->getResult()->getStatus());
+        $this->assertSame('The maximum allowed attempts of 3 was exceeeded.', $messageContext->getResult()->getReason());
+    }
+
+    protected function createMessage(string $job)
+    {
+        $parsedBody = [
+            'class' => [$job, 'execute'],
+            'data' => [],
+        ];
+        $messageBody = json_encode($parsedBody);
+
+        return new NullMessage($messageBody);
+    }
+
+    protected function createMessageResult(QueueMessage $message, ?callable $newMessageAssertions = null)
+    {
+        $queueContext = $this->createMock(Context::class);
+        $consumer = $this->createMock(Consumer::class);
+
+        if ($newMessageAssertions) {
+            $queue = $this->createMock(Queue::class);
+            $consumer->expects($this->once())
+                ->method('getQueue')
+                ->willReturn($queue);
+            $producer = $this->createMock(Producer::class);
+            $producer->expects($this->once())
+                ->method('send')
+                ->with(
+                    $this->equalTo($queue),
+                    $this->callback($newMessageAssertions)
+                );
+            $queueContext->expects($this->once())
+                ->method('createProducer')
+                ->willReturn($producer);
+        } else {
+            $queueContext->expects($this->never())
+                ->method('createProducer');
+        }
```

Comment:
Any reason to not use a real queue & consumer here? Mocks have the tendency to mask failures or cause failures when the code is actually working. We have a few other consumer based tests that you could use as inspiration.

Reply:
No specific reason other than trying to isolate from dependencies. But I understand your point, I believe I can use `QueueManager` to generate what I need which I think is the more the established pattern.

Reply:
Doesn't using mocks make you depend on both the interfaces of these classes and the mocking library as well? :shrug:

Reply:
The interfaces yes, but not the behavior of the classes that implement them so the test will only test a single isolated piece. I updated to use `QueueManager` directly instead. The test does read better this way I think for that's worth.